File: python/src/store/vector.py
"""Vector store for semantic knowledge search.

MVP: Uses SQLite FTS5 for full-text search.
Upgrade path: Replace with ChromaDB or another vector DB
when embedding infrastructure is available.
"""

import re
import sqlite3
from dataclasses import dataclass

from .scope import resolve_db_path
from .types import KnowledgeScope

FTS_SCHEMA = """
CREATE VIRTUAL TABLE IF NOT EXISTS knowledge_fts USING fts5(
  id UNINDEXED,
  content,
  tags
);
"""


@dataclass
class SearchResult:
    id: str
    content: str
    tags: list[str]
    score: float


class VectorStore:
    """Vector store backed by SQLite FTS5."""

    def __init__(self, scope: KnowledgeScope, project_root: str | None = None) -> None:
        db_path = resolve_db_path(scope, project_root)
        self.db = sqlite3.connect(db_path)
        self.db.row_factory = sqlite3.Row
        self.db.execute("PRAGMA journal_mode = WAL")
        self.db.executescript(FTS_SCHEMA)

    def index(self, chunk_id: str, content: str, tags: list[str]) -> None:
        """Index a knowledge chunk for search."""
        self.db.execute(
            "INSERT OR REPLACE INTO knowledge_fts (id, content, tags) VALUES (?, ?, ?)",
            (chunk_id, content, " ".join(tags)),
        )
        self.db.commit()

    def search(self, query: str, limit: int = 5) -> list[SearchResult]:
        """Search by query string using FTS5 ranking."""
        sanitized = _sanitize_fts_query(query)
        if not sanitized:
            return []

        cursor = self.db.execute(
            """SELECT id, content, tags, rank
               FROM knowledge_fts
               WHERE knowledge_fts MATCH ?
               ORDER BY rank
               LIMIT ?""",
            (sanitized, limit),
        )
        rows = cursor.fetchall()

        return [
            SearchResult(
                id=row["id"],
                content=row["content"],
                tags=[t for t in row["tags"].split(" ") if t],
                score=-row["rank"],  # FTS5 rank is negative (lower = better)
            )
            for row in rows
        ]

    def remove(self, chunk_id: str) -> None:
        """Remove an entry from the search index."""
        self.db.execute("DELETE FROM knowledge_fts WHERE id = ?", (chunk_id,))
        self.db.commit()

    def close(self) -> None:
        """Close the database connection."""
        self.db.close()


def _sanitize_fts_query(query: str) -> str:
    """Sanitize query for FTS5 MATCH syntax.

    Splits into tokens and joins with OR for broad matching.
    """
    # Remove non-alphanumeric, non-whitespace characters (Unicode-aware)
    cleaned = re.sub(r"[^\w\s]", " ", query, flags=re.UNICODE)
    tokens = [t for t in cleaned.split() if len(t) > 0]

    if not tokens:
        return ""

    # Join tokens with OR for broad matching
    return " OR ".join(f'"{t}"' for t in tokens)

File: dart/lib/store/vector.dart
import 'package:sqlite3/sqlite3.dart';
import 'scope.dart';
import 'types.dart';

const _ftsSchema = '''
CREATE VIRTUAL TABLE IF NOT EXISTS knowledge_fts USING fts5(
  id UNINDEXED,
  content,
  tags
);
''';

/// Search result from vector/FTS search.
class SearchResult {
  final String id;
  final String content;
  final List<String> tags;
  final double score;

  const SearchResult({
    required this.id,
    required this.content,
    required this.tags,
    required this.score,
  });
}

/// Vector store for semantic knowledge search.
///
/// MVP: Uses SQLite FTS5 for full-text search.
/// Upgrade path: Replace with ChromaDB or another vector DB
/// when embedding infrastructure is available.
class VectorStore {
  final Database _db;

  VectorStore(KnowledgeScope scope, {String? projectRoot})
      : _db = sqlite3.open(resolveDbPath(scope, projectRoot: projectRoot)) {
    _db.execute('PRAGMA journal_mode = WAL');
    _db.execute(_ftsSchema);
  }

  /// Index a knowledge chunk for search.
  void index(String id, String content, List<String> tags) {
    final stmt = _db.prepare(
      'INSERT OR REPLACE INTO knowledge_fts (id, content, tags) VALUES (?, ?, ?)',
    );
    stmt.execute([id, content, tags.join(' ')]);
    stmt.dispose();
  }

  /// Search by query string using FTS5 ranking.
  List<SearchResult> search(String query, {int limit = 5}) {
    final sanitized = _sanitizeFtsQuery(query);
    if (sanitized.isEmpty) return [];

    final results = _db.select(
      '''SELECT id, content, tags, rank
         FROM knowledge_fts
         WHERE knowledge_fts MATCH ?
         ORDER BY rank
         LIMIT ?''',
      [sanitized, limit],
    );

    return results.map((row) {
      final tagsStr = row['tags'] as String;
      return SearchResult(
        id: row['id'] as String,
        content: row['content'] as String,
        tags: tagsStr.split(' ').where((t) => t.isNotEmpty).toList(),
        score: -(row['rank'] as num).toDouble(), // FTS5 rank is negative (lower = better)
      );
    }).toList();
  }

  /// Remove an entry from the search index.
  void remove(String id) {
    final stmt = _db.prepare('DELETE FROM knowledge_fts WHERE id = ?');
    stmt.execute([id]);
    stmt.dispose();
  }

  /// Close the database connection.
  void close() {
    _db.dispose();
  }
}

/// Sanitize query for FTS5 MATCH syntax.
/// Splits into tokens and joins with OR for broad matching.
String _sanitizeFtsQuery(String query) {
  final tokens = query
      .replaceAll(RegExp(r'[^\p{L}\p{N}\s]', unicode: true), ' ')
      .split(RegExp(r'\s+'))
      .where((t) => t.isNotEmpty)
      .toList();

  if (tokens.isEmpty) return '';

  // Join tokens with OR for broad matching
  return tokens.map((t) => '"$t"').join(' OR ');
}
